src/status_im/native_module/impl/module.cljs

```diff
@@ -126,79 +127,57 @@
-
-;; TODO(rasom): temporal solution, should be fixed on status-go side
-(def check-raw-calls-interval 400)
-(def interval-between-calls 100)
-;; contains all calls to jail before with duplicates
-(def raw-jail-calls (atom '()))
-;; contains only calls that passed duplication check
-(def jail-calls (atom '()))
-
-(defn remove-duplicate-calls
-  "Removes duplicates by [jail path] keys, remains the last one."
-  [[all-keys calls] {:keys [jail-id path] :as call}]
-  (if (and (contains? all-keys [jail-id path])
-           (not (#{:subscription :preview} (last path))))
-    [all-keys calls]
-    [(conj all-keys [jail-id path])
-     (conj calls call)]))
-
-(defn check-raw-calls-loop!
-  "Only the last call with [jail path] key is added to jail-calls list
-  from raw-jail-calls"
-  []
-  (go-loop [_ nil]
-    (let [[_ new-calls] (reduce remove-duplicate-calls [#{} '()] @raw-jail-calls)]
-      (reset! raw-jail-calls '())
-      (swap! jail-calls (fn [old-calls]
-                          (concat new-calls old-calls))))
-    (recur (<! (timeout check-raw-calls-interval)))))
-
-(defn execute-calls-loop!
-  "Calls to jail are executed ne by one with interval-between-calls,
-  which reduces chances of response shuffling"
-  []
-  (go-loop [_ nil]
-    (let [next-call (first @jail-calls)]
-      (swap! jail-calls rest)
-      (when next-call
-        (execute-call next-call)))
-    (recur (<! (timeout interval-between-calls)))))
-
-(check-raw-calls-loop!)
-(execute-calls-loop!)
+     #(do
+        (log/debug :call-jail :jail-id jail-id)
+        (log/debug :call-jail :path path)
+        ;; this debug message can contain sensitive info
+        #_(log/debug :call-jail :params params)
+        (let [params' (update params :context assoc
+                              :debug js/goog.DEBUG
+                              :locale rn-dependencies/i18n.locale)
+              cb      (fn [r]
+                        (let [{:keys [result] :as r'} (types/json->clj r)
+                              {:keys [messages]} result]
+                          (log/debug r')
+                          (doseq [{:keys [type message]} messages]
+                            (log/debug (str "VM console(" type ") - " message)))
+                          (callback r')))]
+          (.callJail status jail-id (types/clj->json path) (types/clj->json params') cb))))))
+
+(def ^:private queue-flush-time 400)
+
+(def ^:private call-queue (async/chan))
+(def ^:private deduplicated-calls (async/chan))
+
+(async-util/chunked-pipe! call-queue deduplicated-calls queue-flush-time)
+
+(defn- compare-calls
+  "Used as comparator deciding which calls should be de-duplicated.
+  Whenever we fetch suggestions, we only want to issue the last call
+  done in the `queue-flush-time` window, for all other calls, we have
+  to consider `:params` key as well."
+  [{:keys [jail-id path params]}]
+  (if (= :suggestions (last path))
```

Comment:
`:preview` should be there too

Reply:
@rasom Why ? Are we ever rapidly requesting previews for the same path in succession ?

Reply:
This can happen if user receives more than one command at once after logging in, for instance. Path can be the same, though messages will be different.

Reply:
@rasom I still don't understand why it's the problem, deduplication based only on the `[jail-id path]` tuple will be only performed for suggestion calls, all other calls (including previews) are deduplicated with taking `params` in consideration as well.

Reply:
<img width="375" alt="screen shot 2017-10-25 at 10 17 59" src="https://user-images.githubusercontent.com/2364994/31987990-d612e7ec-b96d-11e7-90c9-8a83e0f28ddc.png">

just reproduced in `feature/proper-jail-call-deduplication` separate messages, same parameters.

Reply:
So the only thing which is different between `:preview` messages is the callback, I see. 
You are right, rather play it safe and to not do any de-duplication (even with params taken into consideration) except for the `:suggestions` calls.
I will change it and add unit tests for those scenarios as well.